README

                              XScreenSaver

                             a collection of
                            free screen savers
                            for X11 and MacOS

                            By Jamie Zawinski
                             and many others

                     http://www.jwz.org/xscreensaver/

===============================================================================

This is the XScreenSaver source code distribution.  It is strongly recommended
that you install a binary release rather than trying to compile it yourself.
Binaries are available for almost all platforms, including MacOS X.  See the
XScreenSaver web site for details.

To compile on a Unix system with X11:

    ./configure
    make
    make install

To compile on MacOS X:

    Use the included XCode project.  Requires XCode 2.4 and MacOS X 10.4.0
    or newer.

Interested in writing a new screen saver?

    See the README.hacking file.

===============================================================================

XScreenSaver has an extensive manual -- please read it!

===============================================================================

Changes since 5.07:   * New hack, `photopile'.
                      * Rewrote `sonar' and `jigsaw' as OpenGL programs.
                      * Minor tweaks to `maze', `m6502', `hypnowheel', and
                        `timetunnel'.
                      * Savers that load images now obey EXIF rotation tags.
                      * Arrgh, more RANDR noise!  Fixes this time for rotated
                        screens, and for systems where RANDR lies and says the
                        screen size is 0x0.
                      * When the password dialog has timed out or been
                        cancelled, don't pop it right back up a second time.
                      * Password timeouts/cancels don't count as
                        "failed logins".
                      * Retired some of the older, less interesting savers:
                        say goodbye to `bubbles', `critical', `flag', `forest',
                        `glforestfire', `lmorph', `laser', `lightning', `lisa',
                        `lissie', `rotor', `sphere', `spiral', `t3d', `vines',
                        `whirlygig', and `worm'.
                      * Merged `munch' and `mismunch'.
                      * Updated `webcollage' to use twitpics.com as well.
Changes since 5.06:   * Xinerama/RANDR tweaks for old-style multi-screen.
                      * Added bumpy skin and cel shading to `skytentacles'.
                      * `flipflop' can load images onto the tiles.
                      * Fixed the bouncing ball in `stairs'.
                      * Added the missing Utah Teapotahedron to `polyhedra'.
                      * `blitspin' works with color images on OSX now.
                      * Added transparency to `stonerview'.
                      * Improved layout of the preferences dialogs: they
                        should all now be usable even on ridiculously tiny
                        laptop screens.
                      * MacOS preferences text fields now prevent you from
                        entering numbers that are out of range.
                      * Added "Reset to Defaults" button on X11.
                      * Added relevant Wikipedia links to many of the screen
                        saver descriptions.
                      * All hacks support the `-fps' option, not just GL ones.
                      * The `-fps' option now shows CPU load.
Changes since 5.05:   * Xinerama/RANDR fixes: this time for sure.  It should
                        now work to add/remove monitors or resize screens at
                        any time.
                      * New hack, `skytentacles'.
                      * New version of `gleidescope'.
                      * Added the `-log' option to the xscreensaver daemon,
                        since a truly shocking number of Linux users seem to
                        have no idea how to redirect output to a file.
                      * Added `-duration' arg to most image-loading hacks,
                        so that they pick a new image every few minutes.
                      * Added an ATM crash to BSOD.
Changes since 5.04:   * New hacks, `cubicgrid', `hypnowheel', and
                        `lcdscrub' (which isn't really a screen saver).
                      * Updates to `m6502' and `gears'.
                      * Fixed double-buffering problems in `cubestorm' and
                        `noof'.
                      * Better handling of horizontal scroll wheels.
                      * Attempt to work around latest Xinerama braindamage:
                        if the server reports overlapping screens, use the
                        largest non-overlapping rectangles available.
                      * Don't warning about receipt of bogus ClientMessages,
                        since Gnome's just never going to stop sending those.
                      * Worked around MacOS 10.5 perl bug that caused the
                        text-displaying hacks to fail on some systems.
                      * Hopefully fixed font-related System Preferences crashes
                        in MacOS savers.
                      * The recent PAM changes broke the "Caps Lock" warning
                        in the password dialog, the failed login warnings, and
                        syslogging.  Fixed all that.
Changes since 5.03:   * Fixed a possible crash in the unlock dialog (more
                        fallout from the recent PAM changes...)
                      * New hacks, `moebiusgears', `abstractile', and
                        `lockward'.
                      * Rewrote `gears' to use better (involute) gear models,
                        and to be more random.
                      * Minor updates to `cwaves', `voronoi', `deco',
                        `glcells', `rd-bomb', `fireworkx' and `webcollage'.
                      * `pong' can now display the current time as the score.
                      * `xmatrix -mode pipe' works better.
                      * Minor tweaks for compilation on MacOS 10.5.0.
Changes since 5.02:   * New hacks, `cwaves', `glcells', `m6502', and `voronoi'.
                      * Minor fixes to `bsod'.
                      * Fixed possible crash with PAM USB-dongle auth.
                      * Updated `webcollage' to track recent Google Images
                        and Flickr changes.
Changes since 5.01:   * Reworked PAM code to support fingerprint readers, etc.
                      * Ported 'webcollage' to OSX.
                      * Added OSX 10.2 and 10.3 kernel panics to `bsod'.
                      * Fixed a Xinerama crash when changing the screen count.
                      * New blobbier `mirrorblob'.
                      * Minor updates to `lisa', `bsod', `ifs', `hypertorus',
                        `polytopes', `circuit', `endgame', `crackberg',
                        `flipflop', `flipscreen3d', `fliptext', and `carousel'.
                      * Enabled multi-threaded OpenGL on OSX.
Changes since 5.00:   * Backed out recent locale-related changes, since they
                        broke far more things than they fixed.
                      * Fail gracefully with ridiculously small window sizes.
                      * `xflame' and `flag' ignore bitmap option on MacOS.
                      * `speedmine' prefs work on MacOS.
                      * Better explosions in `boxed'.
                      * More dynamic motion in `sproingies'.
                      * More options in `flipflop'.
                      * Minor updates to `topblock'.
                      * Various other minor fixes.
Changes since 4.24:   * Ported to MacOS X! (10.4.0 or newer)
                      * API change: instead of providing a single screenhack()
                        function that does not return, screen savers using the
                        screenhack.h framework must now provide "init" and
                        "draw one frame" functions instead.  All bundled savers
                        have been updated; third-party patches will need work.
                      * All image-loading happens asynchronously.
                      * xscreensaver-getimage-file caches the contents of the
                        image directory for a few hours, so consecutive runs
                        won't have to re-list the whole directory tree.
                      * New hacks, `topblock' and `glschool'.
                      * Removed `xteevee' (superceded by `xanalogtv').
                      * Added variable-sized puzzle pieces to `jigsaw'.
                      * Changes to the defaults and command-line options of
                        many hacks to make the .xml files more consistent.
                      * Reap zombies in `glslideshow' and `carousel'.
                      * `sonar' works without setuid on OSX (dgram icmp).
                      * `xmatrix -mode pipe' displays the text of a subprocess.
                      * `endgame' has higher resolution chess-piece models.
                      * `webcollage' takes a -directory option to get images
                        from a local directory.
                      * The RPM spec file no longer auto-restarts xscreensaver
                        when a new version is installed.  Restart it manually.
Changes since 4.23:   * New versions of `cube21', `glsnake', `celtic'.
                      * Backed out a DPMS-related patch that cause desktop
                        flickering with some X servers.
                      * Fixed startup crash in getgroups() when running setuid.
                      * Default to not displaying stderr on the saver window.
                      * Fixed bad free() in "Documentation" button.
                      * Don't try to run hacks that aren't installed.
                      * Minor fixes to various XML config files and man pages.
Changes since 4.22:   * New hacks, `glhanoi', `cube21', `timetunnel',
                        `juggler3d', and `celtic'.
                      * New versions of `tangram', `webcollage', `hypertorus',
                        `polytopes', and `ripples'.
                      * `sonar' is now quiet about unresolvable hosts.
                      * Minor corrections to BASIC code in `apple2'.
                      * xscreensaver-demo now provides an RPM clue when none
                        of the hacks seem to be installed.
                      * Don't install `ant' by default, since there is some
                        Java tool of that name, which was causing confusion.
                        And also it's boring.
                      * Made screen grabbing work again on MacOS 10.4.2.
                      * No longer prints bogus warnings about ClientMessages

po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from

hop.c

/* -*- Mode: C; tab-width: 4 -*- */

#if !defined( lint ) && !defined( SABER )
static const char sccsid[] = "@(#)hop.c	4.03 97/05/10 xlockmore";

#endif

/*-
 * hop.c - Real Plane Fractals for xlock, the X Window System lockscreen.
 *
 * Copyright (c) 1991 by Patrick J. Naughton.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation.
 *
 * This file is provided AS IS with no warranties of any kind.  The author
 * shall have no liability with respect to the infringement of copyrights,
 * trade secrets or any patents by this file or any part thereof.  In no
 * event will the author be liable for any lost revenue or profits or
 * other special, indirect and consequential damages.
 *
 * Revision History:
 * Changes in xlockmore distribution
 * 24-Jun-97: EJK and RR functions stolen from xmartin2.2
 *            Ed Kubaitis <ejk@ux2.cso.uiuc.edu> ejk functions and xmartin 
 *            Renaldo Recuerdo rr function, generalized exponent version
 *              of the Barry Martin's square root function
 * 10-May-97: Compatible with xscreensaver
 * 27-Jul-95: added Peter de Jong's hop from Scientific American
 *            July 87 p. 111.  Sometimes they are amazing but there are a
 *            few duds (I did not see a pattern in the parameters).
 * 29-Mar-95: changed name from hopalong to hop
 * 09-Dec-94: added Barry Martin's sine hop
 * Changes in original xlock
 * 29-Oct-90: fix bad (int) cast.
 * 29-Jul-90: support for multiple screens.
 * 08-Jul-90: new timing and colors and new algorithm for fractals.
 * 15-Dec-89: Fix for proper skipping of {White,Black}Pixel() in colors.
 * 08-Oct-89: Fixed long standing typo bug in RandomInitHop();
 *	      Fixed bug in memory allocation in init_hop();
 *	      Moved seconds() to an extern.
 *	      Got rid of the % mod since .mod is slow on a sparc.
 * 20-Sep-89: Lint.
 * 31-Aug-88: Forked from xlock.c for modularity.
 * 23-Mar-88: Coded HOPALONG routines from Scientific American Sept. 86 p. 14.
 *            Hopalong was attributed to Barry Martin of Aston University
 *            (Birmingham, England)
 */


#ifdef STANDALONE
#define PROGCLASS            "Hop"
#define HACK_INIT            init_hop
#define HACK_DRAW            draw_hop
#define DEF_DELAY            10000
#define DEF_BATCHCOUNT       1000
#define DEF_CYCLES           2500
#define DEF_NCOLORS          200
#define SMOOTH_COLORS
#include "xlockmore.h"		/* in xscreensaver distribution */
#else /* STANDALONE */
#include "xlock.h"		/* in xlockmore distribution */

#endif /* STANDALONE */

#define DEF_MARTIN "False"
#define DEF_EJK1 "False"
#define DEF_EJK2 "False"
#define DEF_EJK3 "False"
#define DEF_EJK4 "False"
#define DEF_EJK5 "False"
#define DEF_EJK6 "False"
#define DEF_RR "False"
#define DEF_JONG "False"
#define DEF_SINE "False"

static Bool martin;
static Bool ejk1;
static Bool ejk2;
static Bool ejk3;
static Bool ejk4;
static Bool ejk5;
static Bool ejk6;
static Bool rr;
static Bool jong;
static Bool sine;

#ifndef STANDALONE
static XrmOptionDescRec opts[] =
{
	{"-martin", ".hop.martin", XrmoptionNoArg, (caddr_t) "on"},
	{"+martin", ".hop.martin", XrmoptionNoArg, (caddr_t) "off"},
	{"-ejk1", ".hop.ejk1", XrmoptionNoArg, (caddr_t) "on"},
	{"+ejk1", ".hop.ejk1", XrmoptionNoArg, (caddr_t) "off"},
	{"-ejk2", ".hop.ejk2", XrmoptionNoArg, (caddr_t) "on"},
	{"+ejk2", ".hop.ejk2", XrmoptionNoArg, (caddr_t) "off"},
	{"-ejk3", ".hop.ejk3", XrmoptionNoArg, (caddr_t) "on"},
	{"+ejk3", ".hop.ejk3", XrmoptionNoArg, (caddr_t) "off"},
	{"-ejk4", ".hop.ejk4", XrmoptionNoArg, (caddr_t) "on"},
	{"+ejk4", ".hop.ejk4", XrmoptionNoArg, (caddr_t) "off"},
	{"-ejk5", ".hop.ejk5", XrmoptionNoArg, (caddr_t) "on"},
	{"+ejk5", ".hop.ejk5", XrmoptionNoArg, (caddr_t) "off"},
	{"-ejk6", ".hop.ejk6", XrmoptionNoArg, (caddr_t) "on"},
	{"+ejk6", ".hop.ejk6", XrmoptionNoArg, (caddr_t) "off"},
	{"-rr", ".hop.rr", XrmoptionNoArg, (caddr_t) "on"},
	{"+rr", ".hop.rr", XrmoptionNoArg, (caddr_t) "off"},
	{"-jong", ".hop.jong", XrmoptionNoArg, (caddr_t) "on"},
	{"+jong", ".hop.jong", XrmoptionNoArg, (caddr_t) "off"},
	{"-sine", ".hop.sine", XrmoptionNoArg, (caddr_t) "on"},
	{"+sine", ".hop.sine", XrmoptionNoArg, (caddr_t) "off"}
};
static argtype vars[] =
{
	{(caddr_t *) & martin, "martin", "Martin", DEF_MARTIN, t_Bool},
	{(caddr_t *) & ejk1, "ejk1", "EJK1", DEF_EJK1, t_Bool},
	{(caddr_t *) & ejk2, "ejk2", "EJK2", DEF_EJK2, t_Bool},
	{(caddr_t *) & ejk3, "ejk3", "EJK3", DEF_EJK3, t_Bool},
	{(caddr_t *) & ejk4, "ejk4", "EJK4", DEF_EJK4, t_Bool},
	{(caddr_t *) & ejk5, "ejk5", "EJK5", DEF_EJK5, t_Bool},
	{(caddr_t *) & ejk6, "ejk6", "EJK6", DEF_EJK6, t_Bool},
	{(caddr_t *) & rr, "rr", "RR", DEF_RR, t_Bool},
	{(caddr_t *) & jong, "jong", "Jong", DEF_JONG, t_Bool},
	{(caddr_t *) & sine, "sine", "Sine", DEF_SINE, t_Bool}
};
static OptionStruct desc[] =
{
	{"-/+martin", "turn on/off sqrt format"},
	{"-/+ejk1", "turn on/off ejk1 format"},
	{"-/+ejk2", "turn on/off ejk2 format"},
	{"-/+ejk3", "turn on/off ejk3 format"},
	{"-/+ejk4", "turn on/off ejk4 format"},
	{"-/+ejk5", "turn on/off ejk5 format"},
	{"-/+ejk6", "turn on/off ejk6 format"},
	{"-/+rr", "turn on/off rr format"},
	{"-/+jong", "turn on/off jong format"},
	{"-/+sine", "turn on/off sine format"}
};

ModeSpecOpt hop_opts =
{20, opts, 10, vars, desc};

#endif /* !STANDALONE */

#define MARTIN 0
#define EJK1 1
#define EJK2 2
#define EJK3 3
#define EJK4 4
#define EJK5 5
#define EJK6 6
#define RR 7
#define CP1 8
#define JONG 9
#define SINE 10
#ifdef OFFENDING
#define OPS SINE			/* Sine might be too close to a Swastika for some... */
#else
#define OPS (SINE+1) 
#endif

typedef struct {
	int         centerx;
	int         centery;	/* center of the screen */
	double      a;
	double      b;
	double      c;
	double      d;
	double      i;
	double      j;		/* hopalong parameters */
	int         inc;
	int         pix;
	int         op;
	int         count;
	int         bufsize;
  XPoint *pointBuffer;	/* pointer for XDrawPoints */
} hopstruct;

static hopstruct *hops = NULL;

void
init_hop(ModeInfo * mi)
{
	Display    *display = MI_DISPLAY(mi);
	GC          gc = MI_GC(mi);
	hopstruct  *hp;
	double      range;

	if (hops == NULL) {
		if ((hops = (hopstruct *) calloc(MI_NUM_SCREENS(mi),
						 sizeof (hopstruct))) == NULL)
			return;
	}
	hp = &hops[MI_SCREEN(mi)];

	hp->centerx = MI_WIN_WIDTH(mi) / 2;
	hp->centery = MI_WIN_HEIGHT(mi) / 2;
	/* Make the other operations less common since they are less interesting */
	if (MI_WIN_IS_FULLRANDOM(mi)) {
			hp->op = NRAND(OPS);
	} else {
		if (martin)
			hp->op = MARTIN;
		else if (ejk1)
			hp->op = EJK1;
		else if (ejk2)
			hp->op = EJK2;
		else if (ejk3)
			hp->op = EJK3;
		else if (ejk4)
			hp->op = EJK4;
		else if (ejk5)
			hp->op = EJK5;
		else if (ejk6)
			hp->op = EJK6;
		else if (rr)
			hp->op = RR;
		else if (jong)
			hp->op = JONG;
		else if (sine)
			hp->op = SINE;
		else
			hp->op = NRAND(OPS);
	}

	switch (hp->op) {
		case MARTIN:
			range = sqrt((double) hp->centerx * hp->centerx +
				     (double) hp->centery * hp->centery) / (10.0 + NRAND(10));

			hp->a = (LRAND() / MAXRAND) * range - range / 2.0;
			hp->b = (LRAND() / MAXRAND) * range - range / 2.0;
			hp->c = (LRAND() / MAXRAND) * range - range / 2.0;
			if (LRAND() & 1)
				hp->c = 0.0;
			/* hp->a = (LRAND() / MAXRAND) * 1500.0 + 40.0;
			hp->b = (LRAND() / MAXRAND) * 17.0 + 3.0;
			hp->c = (LRAND() / MAXRAND) * 3000.0 + 100.0; */
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "sqrt a=%g, b=%g, c=%g\n", hp->a, hp->b, hp->c);
			break;
		case EJK1:
			hp->a = (LRAND() / MAXRAND) * 500.0;
			hp->b = (LRAND() / MAXRAND) * 0.4;
			hp->c = (LRAND() / MAXRAND) * 100.0 + 10.0;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "ejk1 a=%g, b=%g, c=%g\n", hp->a, hp->b, hp->c);
			break;
		case EJK2:
			hp->a = (LRAND() / MAXRAND) * 500.0;
			hp->b = pow(10.0, 6.0 + (LRAND() / MAXRAND) * 24.0);
      if (LRAND() & 1)
        hp->b = hp->b;
			hp->c = pow(10.0, (LRAND() / MAXRAND) * 9.0);
      if (LRAND() & 1)
        hp->c = hp->c;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "ejk2 a=%g, b=%g, c=%g\n", hp->a, hp->b, hp->c);
		case EJK3:
			hp->a = (LRAND() / MAXRAND) * 500.0;
			hp->b = (LRAND() / MAXRAND) * 0.35 + 0.5;
			hp->c = (LRAND() / MAXRAND) * 80.0 + 30.0;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "ejk3 a=%g, b=%g, c=%g\n", hp->a, hp->b, hp->c);
			break;
		case EJK4:
			hp->a = (LRAND() / MAXRAND) * 1000.0;
			hp->b = (LRAND() / MAXRAND) * 9.0 + 1.0;
			hp->c = (LRAND() / MAXRAND) * 40.0 + 30.0;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "ejk4 a=%g, b=%g, c=%g\n", hp->a, hp->b, hp->c);
			break;
		case EJK5:
			hp->a = (LRAND() / MAXRAND) * 600.0;
			hp->b = (LRAND() / MAXRAND) * 0.3 + 0.1;
			hp->c = (LRAND() / MAXRAND) * 90.0 + 20.0;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "ejk5 a=%g, b=%g, c=%g\n", hp->a, hp->b, hp->c);
			break;
		case EJK6:
			hp->a = (LRAND() / MAXRAND) * 100.0 + 550.0;
			hp->b = (LRAND() / MAXRAND) + 0.5;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "ejk6 a=%g, b=%g\n", hp->a, hp->b);
			break;
		case RR:
			hp->a = (LRAND() / MAXRAND) * 100.0;
			hp->b = (LRAND() / MAXRAND) * 20.0;
			hp->c = (LRAND() / MAXRAND) * 200.0;
			hp->d = (LRAND() / MAXRAND) * 0.9;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "rr a=%g, b=%g, c=%g, d=%g\n",
					hp->a, hp->b, hp->c, hp->d);
			break;
		case JONG:
			hp->a = (LRAND() / MAXRAND) * 2.0 * M_PI - M_PI;
			hp->b = (LRAND() / MAXRAND) * 2.0 * M_PI - M_PI;
			hp->c = (LRAND() / MAXRAND) * 2.0 * M_PI - M_PI;
			hp->d = (LRAND() / MAXRAND) * 2.0 * M_PI - M_PI;
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "jong a=%g, b=%g, c=%g, d=%g\n",
					hp->a, hp->b, hp->c, hp->d);
			break;
		case SINE:  /* MARTIN2 */
			hp->a = M_PI + ((LRAND() / MAXRAND) * 2.0 - 1.0) * 0.7;
			/* hp->a = (LRAND() / MAXRAND) * 0.14 + MI_PI - .07; */
  		if (MI_WIN_IS_VERBOSE(mi))
    		(void) fprintf(stdout, "sine a=%g\n", hp->a);
			break;
	}
	if (MI_NPIXELS(mi) > 2)
		hp->pix = NRAND(MI_NPIXELS(mi));
	hp->i = hp->j = 0.0;
	hp->inc = (int) ((LRAND() / MAXRAND) * 200) - 100;
	hp->bufsize = MI_BATCHCOUNT(mi);

	if (!hp->pointBuffer)
		hp->pointBuffer = (XPoint *) malloc(hp->bufsize * sizeof (XPoint));

	XClearWindow(display, MI_WINDOW(mi));

	XSetForeground(display, gc, MI_WIN_WHITE_PIXEL(mi));
	hp->count = 0;
}


void
draw_hop(ModeInfo * mi)
{
	hopstruct  *hp = &hops[MI_SCREEN(mi)];
	double      oldj, oldi;
	XPoint     *xp = hp->pointBuffer;
	int         k = hp->bufsize;

	hp->inc++;
	if (MI_NPIXELS(mi) > 2) {
		XSetForeground(MI_DISPLAY(mi), MI_GC(mi), MI_PIXEL(mi, hp->pix));
		if (++hp->pix >= MI_NPIXELS(mi))
			hp->pix = 0;
	}
	while (k--) {
		oldj = hp->j;
		switch (hp->op) {
			case MARTIN: /* SQRT, MARTIN1 */
				oldi = hp->i + hp->inc;
				hp->j = hp->a - hp->i;
				hp->i = oldj + ((hp->i < 0)
					   ? sqrt(fabs(hp->b * oldi - hp->c))
					: -sqrt(fabs(hp->b * oldi - hp->c)));
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
			case EJK1:
				 oldi = hp->i + hp->inc;
         hp->j = hp->a - hp->i;
         hp->i = oldj - ((hp->i > 0) ? (hp->b * oldi - hp->c) :
           -(hp->b * oldi - hp->c));
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
			case EJK2:
				 oldi = hp->i + hp->inc;
         hp->j = hp->a - hp->i;
         hp->i = oldj - ((hp->i < 0) ? log(fabs(hp->b * oldi- hp->c)) :
           -log(fabs(hp->b * oldi- hp->c)));
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
			case EJK3:
				 oldi = hp->i + hp->inc;
         hp->j = hp->a - hp->i;
         hp->i = oldj - ((hp->i > 0) ? sin(hp->b * oldi)- hp->c :
           -sin(hp->b * oldi) - hp->c);
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
			case EJK4:
				 oldi = hp->i + hp->inc;
         hp->j = hp->a - hp->i;
         hp->i = oldj - ((hp->i > 0) ? sin(hp->b * oldi)- hp->c :
					 -sqrt(fabs(hp->b*oldi-hp->c)) );
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
			case EJK5:
				 oldi = hp->i + hp->inc;
         hp->j = hp->a - hp->i;
         hp->i = oldj - ((hp->i > 0) ? sin(hp->b * oldi)- hp->c :
					 -(hp->b*oldi-hp->c) );
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
			case EJK6:
				 oldi = hp->i + hp->inc;
         hp->j = hp->a - hp->i;
         hp->i = oldj - asin((hp->b * oldi) - (long) (hp->b * oldi));
				xp->x = (int) (hp->i);
				xp->y = (int) (hp->j);
				break;
			case RR: /* RR1 */
				 oldi = hp->i + hp->inc;
         hp->j = hp->a - hp->i;
         hp->i = oldj - ((hp->i < 0) ? -pow(fabs(hp->b * oldi- hp->c), hp->d) :
					 pow(fabs(hp->b*oldi-hp->c), hp->d));
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
			case JONG:
				if (hp->centerx > 0)
					oldi = hp->i + 4 * hp->inc / hp->centerx;
				else
					oldi = hp->i;
				hp->j = sin(hp->c * hp->i) - cos(hp->d * hp->j);
				hp->i = sin(hp->a * oldj) - cos(hp->b * oldi);
				xp->x = hp->centerx + (int) (hp->centerx * (hp->i + hp->j) / 4.0);
				xp->y = hp->centery - (int) (hp->centery * (hp->i - hp->j) / 4.0);
				break;
			case SINE: /* MARTIN2 */
				oldi = hp->i + hp->inc;
				hp->j = hp->a - hp->i;
				hp->i = oldj - sin(oldi);
				xp->x = hp->centerx + (int) (hp->i + hp->j);
				xp->y = hp->centery - (int) (hp->i - hp->j);
				break;
		}
		xp++;
	}
	XDrawPoints(MI_DISPLAY(mi), MI_WINDOW(mi), MI_GC(mi),
		    hp->pointBuffer, hp->bufsize, CoordModeOrigin);
	if (++hp->count > MI_CYCLES(mi))
		init_hop(mi);
}

void
release_hop(ModeInfo * mi)
{
	if (hops != NULL) {
    int screen;

    for (screen = 0; screen < MI_NUM_SCREENS(mi); screen++) {
      hopstruct *hp = &hops[screen];

			if (hp->pointBuffer)
				(void) free((void *) hp->pointBuffer);
    }

		(void) free((void *) hops);
		hops = NULL;
	}
}

void
refresh_hop(ModeInfo * mi)
{
	XClearWindow(MI_DISPLAY(mi), MI_WINDOW(mi));
}
